### PR TITLE
feat(sdk-generator): turn x-local-timezone mandatory for utils.Date

### PR DESCRIPTION
**Pull request by cpaulve-1A**
## Proposed change

Use `x-local-timezone` as an opt-in mechanism to use utils.Date and utils.DateTime .

Set sdk `isStringified` option for dates as true by default

## Related issues

- :rocket: Feature #1808

### CODE REVIEW

File: packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java [hunk 1]
```diff
@@ -512,11 +506,20 @@ public void postProcessModelProperty(CodegenModel model, CodegenProperty propert
         + "list container.");
     }
 
-    // If the x-date-timezone is present in the specification, replace utils.Date with Date
-    if (("utils.Date".equals(property.dataType) || "utils.DateTime".equals(property.dataType)) && property.vendorExtensions.containsKey("x-date-timezone")) {
-      property.dataType = "Date";
-      property.datatypeWithEnum = "Date";
-      property.baseType = "Date";
+    if (property.vendorExtensions.containsKey("x-date-timezone")) {
+      throw new IllegalArgumentException("'x-date-timezone' is deprecated and conflicts with the 'x-local-timezone' vendor." +
```
Comment (vscaiceanu-1a) [inline packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java:510]: Shouldn't we throw if both `x-date-timezone` and `x-local-timezone` are present?
And maybe log a non-blocking warning if `x-date-timezone` is there?

File: packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/readme.md [hunk 2]
```diff
@@ -140,6 +139,8 @@ Simple example:
     properties:
       departureDateTime:
         type: string
+        x-local-timezone:
+        description: If this vendor extension is present send dates without their timezone
```
Comment (vscaiceanu-1a) [inline packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/readme.md:143]: If the `description` is for `x-local-timezone` I think we need an extra tab

File: packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/readme.md [hunk 3]
```diff
@@ -197,23 +198,3 @@ export * from './flight/index';
 
 You can now use departureDateTimeConsideringTimezone to access the timezone information.
 See [utils.Date](https://github.com/AmadeusITGroup/otter/blob/main/packages/%40ama-sdk/core/src/fwk/date.ts) for more information.
-
-### How to keep the timezone (prevent Date replacement with utils.Date)
```
Comment (vscaiceanu-1a) [inline packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/readme.md:201]: I think there's a section in `packages/@ama-sdk/showcase-sdk/readme.md` that needs to be updated as well.

File: packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java (hunk 1 above)
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java:510]: I think we were to completely drop the x-date-timezone once we started supporting the x-local-timezone.

File: packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/readme.md (hunk 2 above)
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/readme.md:143]: Nope, it is for the departure date time object

File: packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java (hunk 1 above)
Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/schematics/typescript/core/openapi-codegen-typescript/src/main/java/com/amadeus/codegen/ts/AbstractTypeScriptClientCodegen.java:510]: As discussed, we need the information of the x-vendor used in the properties in order to set the property original value. This is something we cannot really do just like this (or I guess we would need to rework all the properties as soon as we get the info? This does not seem such a good idea and I am not sure we can pull it off).

I followed your advice and put the migration guide in the error message to help our user to migrate.

File: migration-guides/11.0.md [hunk 4]
```diff
@@ -10,3 +10,44 @@
...
 - `NodePackageNgAddTask` has been removed. `setupDependencies` should be used instead
+
+- The [`stringifyDate` cli option](https://www.npmjs.com/package/@ama-sdk/schematics#dates) is now enabled by default. If 
+you want to use the `Date` format, you will need to set it to `false` via the command line:
+
+```shell
+# with yarn package manager
+yarn run schematics @ama-sdk/schematics:typescript-core --global-property stringifyDate=false
+# with npm package manager
+npm run schematics @ama-sdk/schematics:typescript-core --global-property stringifyDate=false
```
Comment (kpanot) [inline migration-guides/11.0.md:21]:
```suggestion
npm exec schematics @ama-sdk/schematics:typescript-core -- --global-property stringifyDate=false
```

File: packages/@ama-sdk/schematics/project.json [hunk 5]
```diff
@@ -37,7 +37,25 @@
       "executor": "nx:run-script",
       "options": {
         "script": "build:builders"
-      }
+      },
+      "outputs": [
```
Comment (kpanot) [inline packages/@ama-sdk/schematics/project.json:41]: What is the difference with the default one from `nx.json`?
If there is one, can not it be updated to the `nx.json` list?

Comment (cpaulve-1A) [reply packages/@ama-sdk/schematics/project.json:41]: We add the dist/src which contains the schematics interface, I don't think this is something we want on non schematics projects.

File: packages/@ama-sdk/schematics/README.md [hunk 6]
```diff
@@ -127,16 +127,15 @@ Please note that revivers are generated for SDKs that use:
 
 If your specification file includes dates, there are multiple options for the generation of your SDK involving the global property option `stringifyDate`:
 
-- By default, the option `stringifyDate` is false so dates will be generated as either `utils.Date`, `utils.DateTime`, or `Date`.
+- By default, the option `stringifyDate` is set to `true`. Set it to `false` if you want dates will be generated as `Date`.
```
Comment (fpaul-1A) [inline packages/@ama-sdk/schematics/README.md:130]:
```suggestion
- By default, the option `stringifyDate` is set to `true`. Set it to `false` if you want dates to be generated as `Date`.
```

**Review by fpaul-1A — APPROVED** [review]

File: packages/@ama-sdk/showcase-sdk/readme.md [hunk 7]
```diff
@@ -130,12 +130,11 @@ You will need to compute this information with the two timezones -- the airport'
 
 ### Solution proposed to remove the timezone: utils.DateTime
 
-We have introduced the utils.Date object to replace the Date implementation and ignore the timezone.
-As we need to get rid of the timezones more often than not, this will be our default behavior.
-By default, the Date models are replaced with utils.Date.
+The Otter framework has introduced the `utils.Date` object to replace the `Date` implementation and ignore the timezone.
```
Comment (kpanot) [inline packages/@ama-sdk/showcase-sdk/readme.md:133]:
```suggestion
The Otter framework has introduced the `utils.Date` object to replace the `Date` implementation and ignore the timezone of the browser to considered only the one specified in the date returned by the API.
```

File: migration-guides/11.0.md [hunk 8]
```diff
@@ -10,3 +10,44 @@
...
+
+- Prior version 11, the default type for dates was the `utils.Date` and `utils.DateTime` which take the date
+from the API response and remove the timezone to use the local one from the user's machine.
+If a date needed to be expressed in its original timezone, the date specification needed an extra `x-date-timezone` vendor.
+From version 11, the default behaviour is now to keep the timezone of the API response and to use `string` or `Date`
+depending on the `stringifyDate` value. The removal the timezone is now an 'opt-in' mechanism and dates with a timezone 
+that should be ignored needs to add the `x-local-timezone` vendor.
+
+> [!NOTE]
+> You can revert to the previous behaviour thanks to the global property `useLegacyDateExtension`
```
Comment (matthieu-crouzet) [inline migration-guides/11.0.md:53]: Not mandatory
```suggestion
> You can revert to the previous behavior thanks to the global property `useLegacyDateExtension`
```

File: packages/@ama-sdk/create/src/index.it.spec.ts [hunk 9]
```diff
@@ -93,7 +97,7 @@ describe('Create new sdk command', () => {
       }, execAppOptions)
     ).not.toThrow();
     expect(() => packageManagerRun({script: 'build'}, { ...execAppOptions, cwd: sdkPackagePath })).not.toThrow();
-    expect(existsSync(path.join(sdkPackagePath, 'src', 'models', 'base', 'pet', 'pet.reviver.ts'))).toBeTruthy();
+    expect(existsSync(path.join(sdkPackagePath, 'src', 'models', 'base', 'pet', 'pet.ts'))).toBeTruthy();
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/create/src/index.it.spec.ts:100]:
```suggestion
    expect(existsSync(path.join(sdkPackagePath, 'src', 'models', 'base', 'pet', 'pet.ts'))).toBeTruthy();
    expect(existsSync(path.join(sdkPackagePath, 'src', 'models', 'base', 'pet', 'pet.reviver.ts'))).toBeFalsy();
```

[the remark this answers is not in the document]
Comment (cpaulve-1A) [reply packages/@ama-sdk/create/src/index.it.spec.ts:100]: I will keep the other line as it ensures some files have been generated

File: migration-guides/11.0.md [hunk 10]
```diff
@@ -10,3 +10,44 @@
 ## @o3r/schematics
 
 - `NodePackageNgAddTask` has been removed. `setupDependencies` should be used instead
+
+- The [`stringifyDate` cli option](https://www.npmjs.com/package/@ama-sdk/schematics#dates) is now enabled by default. If
```
Comment (sdo-1A) [inline migration-guides/11.0.md:14]:
```suggestion
- The [`stringifyDate` CLI option](https://www.npmjs.com/package/@ama-sdk/schematics#dates) is now enabled by default. If 
```

File: migration-guides/11.0.md [hunk 11]
```diff
@@ -10,3 +10,44 @@
...
+        "globalProperty": {
+          "stringifyDate": false
+        }
+      }
+    }
+  }
+}
+```
+
+- Prior version 11, the default type for dates was the `utils.Date` and `utils.DateTime` which take the date
```
Comment (sdo-1A) [inline migration-guides/11.0.md:45]:
```suggestion
- Prior to version 11, the default type for dates was `utils.Date` and `utils.DateTime` which take the date
```

File: migration-guides/11.0.md [hunk 12]
```diff
@@ -10,3 +10,44 @@
...
+from the API response and remove the timezone to use the local one from the user's machine.
+If a date needed to be expressed in its original timezone, the date specification needed an extra `x-date-timezone` vendor.
```
Comment (sdo-1A) [inline migration-guides/11.0.md:47]:
```suggestion
If a date had to be expressed in its original timezone, the date specification required an extra `x-date-timezone` vendor.
```

File: migration-guides/11.0.md [hunk 13]
```diff
@@ -10,3 +10,44 @@
...
+From version 11, the default behavior is now to keep the timezone of the API response and to use `string` or `Date`
```
Comment (sdo-1A) [inline migration-guides/11.0.md:48]:
```suggestion
As of version 11, the default behavior is now to keep the timezone of the API response and to use `string` or `Date`
```

File: migration-guides/11.0.md [hunk 14]
```diff
@@ -10,3 +10,44 @@
...
+depending on the `stringifyDate` value. The removal the timezone is now an 'opt-in' mechanism and dates with a timezone 
+that should be ignored needs to add the `x-local-timezone` vendor.
```
Comment (sdo-1A) [inline migration-guides/11.0.md:50]:
```suggestion
that should be ignored need to add the `x-local-timezone` vendor.
```

File: packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/readme.md [hunk 15]
```diff
@@ -140,6 +144,8 @@ Simple example:
     properties:
       departureDateTime:
         type: string
+        x-local-timezone:
```
Comment (fpaul-1A) [inline packages/@ama-sdk/schematics/schematics/typescript/shell/templates/base/readme.md:147]:
```suggestion
        x-local-timezone: true
```
?

**Review by fpaul-1A — APPROVED** [review]

**Review by fpaul-1A — DISMISSED** [review]

**Review by fpaul-1A — APPROVED** [review]

**Review by kpanot — APPROVED** [review]

**Pull request merged**